bindings/elixir/native/src/ckzg_wrap.c

```diff
@@ -0,0 +1,839 @@
+        return make_error(env, ckzg_atoms.invalid_commitment_length);
+
+    ErlNifBinary proofs;
+    if (!enif_inspect_binary(env, argv[2], &proofs))
+        return make_error(env, ckzg_atoms.proof_not_binary);
+
+    if (proofs.size % BYTES_PER_PROOF != 0) return make_error(env, ckzg_atoms.invalid_proof_length);
+
+    int blobs_len = blobs.size / BYTES_PER_BLOB;
+    int commitments_len = commitments.size / BYTES_PER_COMMITMENT;
+    int proofs_len = proofs.size / BYTES_PER_PROOF;
+    if (commitments_len != blobs_len || commitments_len != proofs_len)
+        return make_error(env, ckzg_atoms.expected_same_array_size);
+
+    KZGSettings *settings;
+    if (!enif_get_resource(env, argv[3], KZGSETTINGS_RES_TYPE, (void **)&settings))
+        return make_error(env, ckzg_atoms.failed_get_settings_resource);
+
+    bool ok;
+    C_KZG_RET ret = verify_blob_kzg_proof_batch(
+        &ok,
+        (Blob *)blobs.data,
+        (Bytes48 *)commitments.data,
+        (Bytes48 *)proofs.data,
+        blobs_len,
+        settings
+    );
+    if (ret != C_KZG_OK) return make_kzg_error(env, ret);
+
+    return make_success(env, ok ? ckzg_atoms.a_true : ckzg_atoms.a_false);
+}
+
+static ERL_NIF_TERM compute_cells_nif(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
+    if (argc != 2) return make_error(env, ckzg_atoms.incorrect_arg_count);
+
+    ErlNifBinary blob;
+    if (!enif_inspect_binary(env, argv[0], &blob))
+        return make_error(env, ckzg_atoms.blob_not_binary);
+
+    if (blob.size != BYTES_PER_BLOB) return make_error(env, ckzg_atoms.invalid_blob_length);
+
+    KZGSettings *settings;
+    if (!enif_get_resource(env, argv[1], KZGSETTINGS_RES_TYPE, (void **)&settings))
+        return make_error(env, ckzg_atoms.failed_get_settings_resource);
+
+    Cell *cells = enif_alloc(BYTES_PER_CELL * CELLS_PER_EXT_BLOB);
+    if (cells == NULL) return make_error(env, ckzg_atoms.out_of_memory);
+
+    C_KZG_RET ret = compute_cells_and_kzg_proofs(cells, NULL, (Blob *)blob.data, settings);
+    if (ret != C_KZG_OK) {
+        enif_free(cells);
+        return make_kzg_error(env, ret);
+    }
+
+    ERL_NIF_TERM *cells_list = enif_alloc(sizeof(ERL_NIF_TERM) * CELLS_PER_EXT_BLOB);
+    if (cells_list == NULL) {
+        enif_free(cells);
+        return make_error(env, ckzg_atoms.out_of_memory);
+    }
+
+    for (int i = 0; i < CELLS_PER_EXT_BLOB; i++) {
+        ERL_NIF_TERM cell_term;
+        unsigned char *cell_bytes = enif_make_new_binary(env, BYTES_PER_CELL, &cell_term);
+        if (cell_bytes == NULL) {
+            enif_free(cells);
+            enif_free(cells_list);
+            return make_error(env, ckzg_atoms.out_of_memory);
+        }
+
+        memcpy(cell_bytes, &cells[i], BYTES_PER_CELL);
+        cells_list[i] = cell_term;
+    }
+
+    ERL_NIF_TERM cells_term = enif_make_list_from_array(env, cells_list, CELLS_PER_EXT_BLOB);
+
+    enif_free(cells);
+    enif_free(cells_list);
+
+    return make_success(env, cells_term);
+}
```

Comment:
This is totally fine as is, but if you're open to using a `goto` here I would support it. Sort of like this:

https://github.com/ethereum/c-kzg-4844/blob/497de8bc4a67661199a71255bb1c4cb08183d9be/src/eip7594/eip7594.c#L490-L535

For a single exit point with cleanup.

Same thing applies to some other functions below.